catalogs/catalog-hadoop/src/main/java/org/apache/gravitino/catalog/hadoop/authentication/UserContext.java

```diff
@@ -0,0 +1,163 @@
+
+import static org.apache.gravitino.catalog.hadoop.SecureHadoopCatalogOperations.GRAVITINO_KEYTAB_FORMAT;
+import static org.apache.gravitino.catalog.hadoop.authentication.AuthenticationConfig.IMPERSONATION_ENABLE_KEY;
+
+import com.google.common.collect.Maps;
+import java.io.Closeable;
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.UndeclaredThrowableException;
+import java.security.PrivilegedActionException;
+import java.security.PrivilegedExceptionAction;
+import java.util.Map;
+import org.apache.gravitino.NameIdentifier;
+import org.apache.gravitino.catalog.hadoop.authentication.AuthenticationConfig.AuthenticationType;
+import org.apache.gravitino.connector.CatalogInfo;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.security.UserGroupInformation;
+
+public abstract class UserContext implements Closeable {
+
+  private static final Map<NameIdentifier, UserContext> userContextMap = Maps.newConcurrentMap();
+
+  private static void addUserContext(NameIdentifier nameIdentifier, UserContext userContext) {
+    userContextMap.put(nameIdentifier, userContext);
+  }
+
+  public static void clearUserContext(NameIdentifier nameIdentifier) {
+    UserContext userContext = userContextMap.remove(nameIdentifier);
+    if (userContext != null) {
+      try {
+        userContext.close();
+      } catch (IOException e) {
+        throw new RuntimeException("Failed to close user context", e);
+      }
+    }
+  }
+
+  public static void cleanAllUserContext() {
+    userContextMap
+        .values()
+        .forEach(
+            userContext -> {
+              try {
+                userContext.close();
+              } catch (IOException e) {
+                throw new RuntimeException("Failed to close user context", e);
+              }
+            });
+    userContextMap.clear();
+  }
+
+  public static UserContext getUserContext(
+      NameIdentifier nameIdentifier,
+      Map<String, String> properties,
+      Configuration configuration,
+      CatalogInfo catalogInfo) {
+    // Try to get the parent user context.
+    NameIdentifier currentNameIdentifier = NameIdentifier.of(nameIdentifier.namespace().levels());
+    UserContext parentContext = null;
+    while (!currentNameIdentifier.namespace().isEmpty()) {
+      if (userContextMap.containsKey(currentNameIdentifier)) {
+        parentContext = userContextMap.get(currentNameIdentifier);
+        break;
+      }
+      currentNameIdentifier = NameIdentifier.of(currentNameIdentifier.namespace().levels());
+    }
+
+    if (configuration == null) {
+      configuration = new Configuration();
+    }
+    AuthenticationConfig authenticationConfig = new AuthenticationConfig(properties);
+
+    // If we do not set the impersonation, we will use the parent context;
+    boolean enableUserImpersonation = false;
+    if (properties.containsKey(IMPERSONATION_ENABLE_KEY)) {
```

Comment:
Does this handle the default value correctly?

Reply:
done